### PR TITLE
[hmac] Initial implementation of save & restore

### PR DESCRIPTION
**Pull request by andreaskurth**
This is an initial implementation of save & restore for HMAC (issue #49), which should allow SW to switch between different parallel message streams. In short, SW usage follows a pattern like this:
1. Start processing message stream A by configuring HMAC and then setting the `CMD.hash_start` bit.
2. Write an arbitrary number of message blocks to HMAC's `MSG_FIFO`.
3. Stop HMAC by setting the `CMD.hash_stop` bit and wait for the `hmac_done` interrupt (or poll the interrupt status register).
4. Save the context by reading the `DIGEST_0`..`7` and `MSG_LENGTH_`{`LOWER`,`UPPER`} registers. (The values in the `CFG` register must also be preserved, but that is purely SW-controlled so doesn't need to be read from HW.)
5. Repeat steps 1-4 for message stream B.
6. Restore the context of message stream A by writing the `CFG`, `DIGEST_0`..`7`, and `MSG_LENGTH_`{`LOWER`,`UPPER`} registers.
7. Continue processing message stream A by setting the `CMD.hash_continue` bit.
8. Write an arbitrary number of message blocks to HMAC's `MSG_FIFO`.
9. Continue this with as many message blocks and parallel message streams as needed. The final hash for any message stream can be obtained at any time (no need for complete blocks) as before by setting `CMD.hash_process` and waiting for the `hmac_done` interrupt / status bit, finally reading the digest from the `DIGEST` registers.

A noteworthy limitation of this implementation is that context can only be switched on complete message blocks (512 bit for SHA-256). This is so that no additional internal state (i.e., working variables and round counter) needs to be exposed to and controlled from SW. As noted in [issue #49](https://github.com/lowRISC/opentitan/issues/49#issuecomment-1935538297), this shouldn't be a major limitation for SW, though.

Although this PR adds preliminary DV support for the changes, save & restore does not yet get checked by the tests (the `save_and_restore_pct` randomization knob is currently 0 for all tests). I have manually verified that save & restore works under the basic operating conditions for HMAC and SHA but this needs to be tested much more thoroughly, especially the corner cases. When one enables save & restores in the tests, one will get errors from DV. My first assessment is that these are DV limitations, not RTL bugs, though.

I ran a full regression on HMAC to check that this doesn't break existing functionality or DV, and I got the same pass rates as reported in nightlies -- so no breakage of existing RTL & DV expected.

Opens before this PR can be merged:
- [x] Update documentation (mainly Programmer's Guide and Theory of Operation).

Opens that can be resolved in a follow-up PR:
- [ ] Extend testplan and coverplan to test and cover save & restore feature.
- [ ] Enable save & restores in tests and make necessary extensions to the scoreboard as well as `burst_wr_msg` so this gets correctly checked and fully covered.
- [ ] Add additional directed test that covers context switching between multiple message streams (not sure if this has to be an UVM test, maybe a SW test would suffice?).

### CODE REVIEW

**Comment by andreaskurth** [comment]
@gdessouky As discussed, here's a first implementation of save & restore. I would be super happy to get your feedback on this. Unless you think this is fundamentally broken, I think it would be good to get this merged soon, so that SW folks can write a small test that exercises this to check that it meets their needs.

**Comment by andreaskurth** [comment]
kokoro reported lint errors that I've fixed by now; the remaining _flow errors_ seem to be spurious.

File: hw/ip/prim/rtl/prim_sha2_32.sv
Comment (rswarbrick) [inline hw/ip/prim/rtl/prim_sha2_32.sv:31]: Doesn't this (first) commit also need the signals to be wired up in `hmac.sv`? Presumably, you can tie `digest_we_i` to ground there.

**Review by rswarbrick — COMMENTED** [review]
Lots of nitty comments, but this looks sensible in general.

In future, I'd definitely suggest splitting up giant PRs into chains that depend on one another. It becomes a bit easier to review and you can "merge the first chunk" rather quicker :-)

File: hw/ip/hmac/data/hmac.hjson
Comment (msfschaffner) [inline hw/ip/hmac/data/hmac.hjson:205]: @moidx @jadephilipoom @ballifatih @gdessouky can you take a look at the API changes to double check this is fulfills the requirements of the cryptolib?

File: hw/ip/hmac/doc/programmers_guide.md
Comment (msfschaffner) [inline hw/ip/hmac/doc/programmers_guide.md:101]: @moidx @jadephilipoom @ballifatih FYI

**Comment by msfschaffner** [comment]
Thanks for doing this (and splitting up the work into multiple easy-to-review commits)!
Overall this looks good to me, but it'd be great to get an ACK from some of the cryptolib devs to make sure this interface fulfills their requirements.

**Review by jadephilipoom — APPROVED** [review]
This looks good to me! I can confirm that the need to only stop at full message blocks shouldn't be an issue for SW (we can just buffer the trailing partial message block and store it along with the hash state until we're ready to resume).

**Comment by gdessouky** [comment]
Thanks @andreaskurth, this looks good to me too, but I still hope to go through this in detail later today. Sorry to not have gotten to it yet, but been struggling with DV issues for the extended digest feature at my end :(

File: hw/ip/prim/rtl/prim_sha2_32.sv
Comment (andreaskurth) [reply hw/ip/prim/rtl/prim_sha2_32.sv:31]: Good catch, thanks! Fixed

File: hw/ip/prim/rtl/prim_sha2.sv
Comment (andreaskurth) [reply hw/ip/prim/rtl/prim_sha2.sv]: Yes, added a comment

File: hw/ip/prim/rtl/prim_sha2_32.sv
Comment (andreaskurth) [reply hw/ip/prim/rtl/prim_sha2_32.sv]: Indeed, fixed

File: hw/ip/hmac/data/hmac.hjson
Comment (andreaskurth) [reply hw/ip/hmac/data/hmac.hjson]: Agreed, and I changed the name of the bit to `hash_continue` (together with the names of all associated signals)

File: hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv
Comment (andreaskurth) [reply hw/ip/hmac/dv/env/seq_lib/hmac_base_vseq.sv]: Hehe, that's what I get from copy-pasting the `trigger_hash` task :slightly_smiling_face: I clarified the comment in both tasks

File: hw/ip/prim/rtl/prim_sha2.sv
Comment (andreaskurth) [reply hw/ip/prim/rtl/prim_sha2.sv]: Yep, fixed

Comment (andreaskurth) [reply hw/ip/prim/rtl/prim_sha2.sv]: Sure, done

File: hw/ip/hmac/dv/tests/hmac_base_test.sv
Comment (andreaskurth) [reply hw/ip/hmac/dv/tests/hmac_base_test.sv]: Agreed and inlined into `build_phase`

File: hw/ip/hmac/doc/programmers_guide.md
Comment (andreaskurth) [reply hw/ip/hmac/doc/programmers_guide.md]: Okay, changed

**Comment by andreaskurth** [comment]
Force-pushed to address @rswarbrick's feedback. Most notably for users and other reviewers, the `hash_restart` command bit is now called `hash_continue`.

File: hw/ip/hmac/data/hmac.hjson
Comment (vogelpi) [inline hw/ip/hmac/data/hmac.hjson:312]: Would it make sense to mention the in the register description that writes by software are ignored unless the SHA core is disabled? Actually, don't you need to update the tags for the automated CSR tests as well?

File: hw/ip/hmac/data/hmac.hjson
Comment (vogelpi) [inline hw/ip/hmac/data/hmac.hjson:334]: I suggest updating these comments to specify that software can also write those and more importantly **when** software can write them.

File: hw/ip/hmac/rtl/hmac.sv
Comment (vogelpi) [inline hw/ip/hmac/rtl/hmac.sv:141]: Nit: process the message

Comment (vogelpi) [reply hw/ip/hmac/rtl/hmac.sv:141]: I guess, `process` here means to stop receiving input, add the padding and then doing the final absorb, right?

File: hw/ip/hmac/rtl/hmac.sv
Comment (vogelpi) [inline hw/ip/hmac/rtl/hmac.sv:166]: Q: Why is it even possible that hash_running is still active, i.e., the core is still processing even after leaving `DoneAwaitBlockHashed`?

File: hw/ip/hmac/doc/programmers_guide.md
Comment (vogelpi) [inline hw/ip/hmac/doc/programmers_guide.md:87]: Would it make sense to add actual references to the register documentation here, i.e., `CFG -> [CFG](registers.md#cfg)`?

**Review by vogelpi — APPROVED** [review]
Thanks for your work on this @andreaskurth ! I've focused my review on the RTL part and this looks very good. I have a few questions and minor details but don't want to block merging this PR.

File: hw/ip/hmac/doc/registers.md
Comment (gdessouky) [inline hw/ip/hmac/doc/registers.md:210]: We will definitely need another pass on the documentation in another PR after both features are merged to consolidate on the references to the digest and key CSRs, etc.

File: hw/ip/hmac/data/hmac.hjson
Comment (gdessouky) [inline hw/ip/hmac/data/hmac.hjson:1]: We would need to also save & restore the `digest_size` and `key_length` CSRs (PR https://github.com/lowRISC/opentitan/pull/21604) and similarly make them  
`swaccess: "rw"`
`hwaccess: "hrw"`

File: hw/ip/hmac/rtl/hmac.sv
Comment (gdessouky) [inline hw/ip/hmac/rtl/hmac.sv:157]: If SW doesn't comply and send in eventually a message with size = multiples of block size, it would get stuck here infinitely and the IP gets locked up, right? 
Do you think we can mandate from the SW to write `reg_hash_stop` only after the message feeding to the FIFO is completed? Then we can check that the message length complies with what we expect, otherwise flag an error.

File: hw/ip/hmac/rtl/hmac.sv
Comment (gdessouky) [reply hw/ip/hmac/rtl/hmac.sv:166]: I believe the FSM transitions to this state from `DoneAwaitBlockHashed` once the message length fed into the FIFO reaches a multiple of the message block size, but now the hashing of that last block actually runs.

File: hw/ip/prim/rtl/prim_sha2.sv
Comment (gdessouky) [reply hw/ip/prim/rtl/prim_sha2.sv:412]: I've understood in an earlier review when merging this primitive that we can drop the begin/end for single line assignments if each statement fits on a single line  https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#begin--end. 

Can always make a pass since my PR also touches this primitive and add the begin/end.

File: hw/ip/hmac/rtl/hmac.sv
Comment (gdessouky) [inline hw/ip/hmac/rtl/hmac.sv:205]: In PR https://github.com/lowRISC/opentitan/pull/21604 I gate updating the digest CSRs with `intr_hmac_done_o` to hold the previous digest value until the current hash computation concludes, and not write intermediate digest changes to the CSRs. This will conflict with this change here.

File: hw/ip/prim/rtl/prim_sha2.sv
Comment (andreaskurth) [reply hw/ip/prim/rtl/prim_sha2.sv:412]: I agree with Ghada that our style guide allows omitting `begin` and `end` if each semicolon-terminated statement fits on a single line

File: hw/ip/hmac/data/hmac.hjson
Comment (andreaskurth) [reply hw/ip/hmac/data/hmac.hjson:312]: Thanks for the suggestion, I extended the description for the digest registers and the message length registers.

I don't think these registers need exclusion tags for the automated CSR tests because they will be readable and writable as expected during the automated CSR tests: `CFG.sha_en` is excluded from writes by the automated CSR tests, so it will remain `0` during these tests.

File: hw/ip/hmac/doc/registers.md
Comment (andreaskurth) [reply hw/ip/hmac/doc/registers.md:210]: Agreed, could you please create an issue to track this?

File: hw/ip/hmac/rtl/hmac.sv
Comment (andreaskurth) [reply hw/ip/hmac/rtl/hmac.sv:157]: SW can set `CMD.hash_stop` to 1 before or after it has written a complete message block. It *must* eventually complete the message block, though. If it doesn't, HMAC never advances beyond this state -- correct. We can create an issue to improve robustness on this, but I wouldn't prioritize it. In the worst case, SW can always write zeros/garbage to the message FIFO until HMAC completes. This will make the current digest useless, but it will unconditionally unblock HMAC.

File: hw/ip/hmac/data/hmac.hjson
Comment (andreaskurth) [reply hw/ip/hmac/data/hmac.hjson]: Thx, done

File: hw/ip/hmac/rtl/hmac.sv
Comment (andreaskurth) [reply hw/ip/hmac/rtl/hmac.sv:166]: The `DoneAwaitBlockHashed` state is entered when the HMAC module receives a stop command and left when the complete message block has been provided. Then the SHA module hashes the message block, during which `hash_running` is high. Once the SHA module is done and it has updated the digest, it lowers `hash_running`.

Your confusion possibly comes from the name `DoneAwaitBlockHashed`. I've now changed that to `DoneAwaitMessageComplete` to hopefully better reflect what's going on.

File: hw/ip/hmac/doc/programmers_guide.md
Comment (andreaskurth) [reply hw/ip/hmac/doc/programmers_guide.md]: Thanks, added

File: hw/ip/hmac/data/hmac.hjson
Comment (vogelpi) [reply hw/ip/hmac/data/hmac.hjson:315]: This makes a lot of sense, thanks for explaining.

File: hw/ip/hmac/rtl/hmac.sv
Comment (vogelpi) [reply hw/ip/hmac/rtl/hmac.sv:166]: Ah, I see. It makes sense now and I think renaming the state for disambiguation is also a good idea.

File: hw/ip/hmac/rtl/hmac.sv
Comment (gdessouky) [reply hw/ip/hmac/rtl/hmac.sv:157]: Yes, agree, definitely not priority at the moment. Will spin out an issue to track it; because one of the motivations for the `save & restore` feature was not to block the HMAC indefinitely by one SW thread and deny access to it from other threads.

**Comment by msfschaffner** [comment]
Awesome, thanks for getting this merged!

File: hw/ip/hmac/doc/programmers_guide.md
Comment (cdgori) [reply hw/ip/hmac/doc/programmers_guide.md:85]: Following along in reverse chronological order, but now peeking ahead - I believe this potential error is tracked in #21710 

(Concur with @andreaskurth that padding with zeros is not a good option - if the error is relatively cheap to implement it is probably more SW-friendly to generate it.)

**Pull request merged**
